src/content/blog/2024-12-20-the-tree-shaking-journey-in-angular-a-deep-dive.mdx

```diff
@@ -0,0 +1,342 @@
+
+```shell
+npx nx generate @nx/angular:library --directory=libs/ng-publishable --publishable=true --importPath=@ngx-santosh/ng-publishable --no-interactive 
+```
+
+Let's add a few components; we will go with standalone components:
+
+```shell
+npx nx generate @nx/angular:component --path=libs/ng-publishable/src/lib/accordion --export=true --no-interactive
+
+npx nx generate @nx/angular:component --path=libs/ng-publishable/src/lib/grid --export=true --no-interactive
+```
+
+Let's publish the library and use `GridComponent` in our Application:
+
+```typescript
+import {GridComponent, AccordionComponent} from '@ngx-santosh/ng-publishable';
+
+@Component({
+ imports: [GridComponent],
+ selector: 'app-root',
+ templateUrl: './app.component.html',
+ styleUrl: './app.component.css',
+})
+export class AppComponent {
+ title = 'hello-world';
+}
+```
+
+And add the Grid component to the template.
+
+```html
+<lib-grid/>
+```
+
+lets run the build and generate the bundle again 
+
+```shell
+nx run hello-world:build:production --sourceMap
+
+npx source-map-explorer dist/apps/hello-world/browser/*.* --html report/hello-world-publish-libs.html  
+```
+
+
+<BlogImage src={publishable} alt="main bundle shows 395 Bytes of code from the GodModuleLibModule" />
+
+
+We still get the tree shaking with publishable libraries. The final code contains only 248 Bytes of grid component code, and once you add more components, you only pay for those components.
+
+
+### Library with secondary-entrypoints
+
+Why should we still read this section and go with secondary entry points even though we have only been able to get the minimal code in the production bundle with standalone components in the exercise so far? 
+
+Remember, once your library becomes too big and has hundreds of imports, which is a common scenario for most developers working on the design system.
+
+If you are not publishing your library on artifactory, I recommend splitting them into smaller libraries; if you are using Nx, you can get the build cache, which will help you improve your build and test times.
+
+But for publishable libraries, where we need to bundle all code into a single package and publish on an artifactory, splitting to smaller libs is not an option; in this case, I recommend going with secondary-entry points.
+
+
+## Some closing notes
+
+1. If you are using Angular 14 or below, the exercise we did may not work to reduce your bundle size, but my recommendation is to start adopting standalone. Once you migrate to Angular 15, use `ng g @angular/core:standalone` to get the minimal code in your final production bundle.
+1. The above suggestion applies to your application and library code.
+1. For libraries with a lot of components, 
+  1. I recommend splitting them into smaller libs.
+  1. If it is publishable, I recommend going with secondary-entrypoint.
+1. With Latest Angular language service you get get warning in IDE if you are importing unused Component
```

Comment:
We can fix the repetition of the word get

Reply:
thanks